File: get_the_offer/2-旋转数组的最小数字/code.rb
def min_number_in_rotate_array(nums)
  left = 0
  right = nums.size - 1

  while left < right
    mid = left + (right - left)/2
    if nums[mid] > nums[right]
      left = mid + 1
    else
      right = mid
    end
  end
  nums[right]
end


# Divide and Conquer
def find_min(nums)

    helper(nums, 0, nums.size - 1);
end

def helper(nums, left, right)
    return nums[left] if nums[left] <= nums[right]

    mid = left + (right - left)/2
    [helper(nums, left, mid), helper(nums, mid+1, right)].min
end

File: get_the_offer/2-旋转数组的最小数字/code.java
public int findMin(int[] nums) {
  int left = 0;
  int right = nums.length - 1;
  while (left < right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] > nums[right]) left = mid + 1;
    else right = mid;
  }
  return nums[right];
}

// Divide and Conquer
int findMin(vector<int>& nums) {
        return helper(nums, 0, (int)nums.size() - 1);
    }
int helper(vector<int>& nums, int start, int end) {
  if (nums[start] <= nums[end]) return nums[start];
  int mid = (start + end) / 2;
  return min(helper(nums, start, mid), helper(nums, mid + 1, end));
}
